Bugfix open workspace assumption

Maxattax97 opened the pull request:
This fixes #50 by allowing a fresh VS Code session with no workspaces open to not halt the language server.

It also fixes the error message by converting the data property into a string (rather than `[object Object]`).

I apologize for my formatting slip up, I pushed to my fork _then_ realized NeoVim auto-formatted the rest of the code... so a bit of commit bloat.

# How to reproduce
1. Setup VS Code so that the next time it opens, there are no workspaces opened.
2. Enter VS Code in debug mode with the extension loaded.
3. Open a JSON file and start writing.
4. Crashes.

# Critical [line](https://github.com/openlawlibrary/pygls/pull/51/commits/820f85a9e42303173b76bef56c54610abf9b6fee#diff-f04f719a2092ca4cc1f1b05b8f7181a3L532) of code
```
        workspace_folders = getattr(params, 'workspaceFolders', [])
        for folder in workspace_folders:
            self.workspace.add_folder(folder)
```

# VS Code Error
```
[Error - 12:11:01 AM] Server initialization failed.
  Message: TypeError: 'NoneType' object is not iterable
  Code: -32602 
[object Object]
```
This is now fixed so it prints a full traceback instead of `[object Object]`.

danixeee commented:
@Maxattax97 thank you for your contribution! I will review it shortly.

Please take a look at [contributing](https://github.com/openlawlibrary/pygls/blob/master/CONTRIBUTING.md) document, section [Final Checklist](https://github.com/openlawlibrary/pygls/blob/master/CONTRIBUTING.md#final-checklist).

Maxattax97 commented:
I've updated my original comment, added your proposed changes, and including documentation where needed in markdown files.

Let me know if there's anything else to be done!

Maxattax97 commented:
Can't say I understand why the CI failed?

danixeee commented on CONTRIBUTORS.md:
```diff
@@ -3,3 +3,4 @@
 - [@augb](https://github.com/augb)
 - [Daniel Elero](https://github.com/danixeee)
 - [yorodm](https://github.com/yorodm)
+- [Max O'Cull](https://github.com/Maxattax97)
```
Contributors should be ordered alphabetically (https://github.com/openlawlibrary/pygls/blob/6e16f41da50ce0dcd83499d2d45adcb8dd10717d/CONTRIBUTORS.md#contributors-alphabetical).

Maxattax97 replied:
Can't believe I missed that :sweat_smile:

danixeee commented on pygls/protocol.py:
```diff
@@ -526,11 +526,8 @@ def bf_initialize(self, params: InitializeParams):
         workspace_folders = getattr(params, 'workspaceFolders', [])
-        for folder in workspace_folders:
-            self.workspace.add_folder(folder)
+        self.workspace = Workspace(root_uri, self, workspace_folders)
```
Workspace constructor changed, so `self` argument have to be removed.

danixeee commented on pygls/workspace.py:
```diff
@@ -128,14 +128,17 @@ def word_at_position(self, position: Position) -> str:
 class Workspace(object):
 
-    def __init__(self, root_uri, lsp):
+    def __init__(self, root_uri, workspace_folders=[]):
```
Can you please set `workspace_folders` to `None`, instead of `[]`?

danixeee commented on tests/test_workspace.py:
```diff
@@ -18,7 +18,7 @@
-from pygls import uris
+from pygls import uris, Workspace
 from pygls.types import TextDocumentItem, WorkspaceFolder
 
```
Should be -> `from pygls.workspace import Workspace`

danixeee commented on tests/test_workspace.py:
```diff
@@ -66,3 +66,12 @@ def test_remove_document(workspace):
     assert workspace.get_document(DOC_URI).source == DOC_TEXT
     workspace.remove_document(DOC_URI)
     assert workspace.get_document(DOC_URI)._source is None
+
```
We are using 2 blank lines between functions.

Maxattax97 replied:
Not sure why my auto formatting didn't catch this.

Maxattax97 commented:
I'm relatively new to Python and still learning some of it's development workflow, thanks for your patience :smile:. I didn't know you could locally install like that (`pip3 install .`) into a venv, good to know from now on.

This is also my first in-depth code review. I've made PR's before, but never had the opportunity to really go through the process, so this is good :+1:.

Maxattax97 commented on examples/json-extension/server/tests/unit/test_features.py:
```diff
@@ -17,7 +17,7 @@
 
 import pytest
-from unittest.mock import Mock
+from mock import Mock
```
The line I removed here works locally for me, but appveyor hates it and needs the added line..?

danixeee replied:
That import is needed because `Mock` is used on several places in that module (see [test_features.py line 41](https://github.com/openlawlibrary/pygls/blob/master/examples/json-extension/server/tests/unit/test_features.py#L41)).

dgreisen commented:
@Maxattax97 Thanks for your contribution. When your OpenSCAD languange server is up, we'd love if you added it to https://github.com/openlawlibrary/pygls/blob/master/Implementations.md.
